sklearn/utils/validation.py

```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
         raise ValueError('`{}`= {}, must be <= {}.'.format(name, x, max_val))
 
 
+def _check_psd_eigenvalues(lambdas, warn_on_zeros=False):
+    """Check the eigenvalues of a positive semidefinite (PSD) matrix.
+
+    Checks the provided array of PSD matrix eigenvalues for numerical or
+    conditioning issues and returns a fixed validated version. This method
+    should typically be used if the PSD matrix is user-provided (e.g. a
+    Gram matrix) or computed using a user-provided dissimilarity metric
+    (e.g. kernel function), or if the decomposition process uses approximation
+    methods (randomized SVD, etc.).
+
+    It checks:
+
+    - that there are no significant imaginary parts in eigenvalues (more than
+      1e-5 times the maximum real part). If this check fails, it raises a
+      ``ValueError``. Otherwise all non-significant imaginary parts that may
+      remain are removed.
+
+    - that eigenvalues are not all negative. If this check fails, it raises a
+      ``ValueError``
```

Comment:
Just a note that we have a somewhat similar issue in GPs and there we just raise a warning if the covariance matrix's diagonal (IIRC) has negative values and simply set them to zero.

Reply:
indeed that seems to be a similar purpose and similar behaviour (warning + zero-setting). Consistency without sync ! :)